src/System.ServiceModel.Federation/src/System/ServiceModel/Federation/WSTrustChannel.cs

```diff
@@ -0,0 +1,427 @@
+        /// The <see cref="TimeSpan" /> that specifies how long the open operation has to complete before timing out.
+        /// </param>
+        void ICommunicationObject.Open(TimeSpan timeout)
+        {
+            RequestChannel.Open(timeout);
+        }
+
+        /// <summary>
+        /// Causes a communication object to transition from the created state into the opened state. 
+        /// </summary>
+        void ICommunicationObject.Open()
+        {
+            RequestChannel.Open();
+        }
+
+        /// <summary>
+        /// Gets the current state of the communication-oriented object.
+        /// </summary>
+        CommunicationState ICommunicationObject.State
+        {
+            get { return RequestChannel.State; }
+        }
+        #endregion
+
+        #region IWSTrustChannelContract Members
+        /// <summary>
+        /// Sends a WS-Trust Issue message to an endpoint STS
+        /// </summary>
+        /// <param name="trustRequest">The <see cref="WsTrustRequest" /> that represents the request to the STS.</param>
+        /// <returns>A <see cref="SecurityToken" /> that represents the token issued by the STS.</returns>
+        public async virtual Task<WCFSecurityToken> IssueAsync(WsTrustRequest trustRequest)
+        {
+            Message requestMessage = CreateRequest(trustRequest);
+            Message response = await Task.Factory.FromAsync(RequestChannel.BeginRequest, RequestChannel.EndRequest, requestMessage, null, TaskCreationOptions.None).ConfigureAwait(false);
+            if (response.IsFault)
+            {
+                MessageFault fault = MessageFault.CreateFault(response, WSTrustChannel.FaultMaxBufferSize);
+                string action = null;
+                if (response.Headers != null)
+                    action = response.Headers.Action;
+
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(FaultException.CreateFault(fault, action));
+            }
+
+            WsTrustResponse trustResponse = TrustSerializer.ReadResponse(response.GetReaderAtBodyContents());
+            WCFSecurityToken token = WSTrustUtilities.CreateGenericXmlSecurityToken(trustRequest, trustResponse, SerializationContext, null);
+
+            if (token == null)
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new SecurityTokenException(SR.Format(SR.TokenProviderUnableToGetToken, string.IsNullOrEmpty(Address) ? ToString() : Address)));
+            
+            return token;
+        }
+        #endregion
+
+        #region IWSTrustContract  Members
+        /// <summary>
+        /// Sends a WS-Trust Cancel message to an endpoint.
+        /// </summary>
+        /// <param name="message">The <see cref="Message" /> that contains the instructions for the request to the STS.</param>
+        /// <returns>The <see cref="Message" /> returned from the STS.</returns>
+        public async Task<Message> CancelAsync(Message message)
+        {
```

Comment:
I don't understand the purpose of these methods. They could be sending anything to the server.

Reply:
Yes the message is open, this was part of the original WSTrustChannel object.
This allows users to send messages for Cancel, Renew, Validte.
I have sample code to show how to use this.